docs/blog/2019-03-03-new-schema-customization/index.md

```diff
@@ -0,0 +1,270 @@
+  const { createTypes } = actions
+  const typeDefs = `
+    # One must say that the type is a Node
+    type AuthorJson implements Node {
+      # However Node fields are optional and you don't have to add them
+      name: String
+      birthday: Date
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+Gatsby will now know that you want a Date and not override it with a string.
+
+You can specify types for some or all of the fields that you have on the given node type. Gatsby will add missing fields. This behaviour can be controlled with `@infer` and `@dontInfer` directives.
+
+```graphql
+# For this type `name` won't be added
+@dontInfer()
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+# For this type `name` won't be added but birtday won't get a Gatsby resolver for date formatting
+@dontInfer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  birthday: Date
+}
+
+
+# For this type all fields `name`. Current default behaviour, but allows one to be explicit about it.
+@infer()
+type AuthorJson implements Node {
+  id: ID!
+}
+
+# Birthday will be Date, but we won't add a Gatsby resolver for date formatting
+@infer(noDefaultResolvers: true)
+type AuthorJson implements Node {
+  id: ID!
+}
+```
+
+## `createResolvers`
+
+This is a similar API to `setFieldsOnNodeType` in that it allows to add new fields and resolvers for them. However, this one is ran last, so you'd have all the schema available to you in it. Also it's possible to extend types like `Query` to add more root resolvers.
+
+```js
+exports.createResolvers = ({ createResolvers, schema }) => {
+  createResolvers({
+    AuthorJson: {
+      // Modify birthday resolver so that it uses 1970-01-01 as default date
+      birthday: {
+        resolve(source, args, context, info) {
+          // original resolver available as "info.originalResolver"
+          if (Number.isNaN(new Date(source['birthday'])) {
+            return info.originalResolver({
+              ...source,
+              birthday: '1970-01-01',
+            }, args, context, info)
+          } else {
+            return info.originalResolver(source, args, context, info)
+          }
+        },
+      },
+    },
+  })
+}
+```
+
+It's also possible to create new root fields, for example one that will return all author names as strings.
+
+```js
+createResolvers({
+  Query: {
+    allAuthorFullNames: {
+      type: `[String!]!`,
+      resolve(source, args, context, info) {
+        const authors = context.nodeModel.getAllNodes(
+          {
+            type: `AuthorJson`,
+          },
+          { path: context.path }
+        )
+        return authors.map(author => author.name)
+      },
+    },
+  },
+})
+```
+
+Notice the `context.nodeModel`. We expose our internal node storage to the resolvers, so that one can fetch data from there. Full documentation of nodeModel is available at TODO.
+
+You can also see TODO ADD using resolvers example.
+
+# Why was it needed?
```

Comment:
I think this section would work better if it was moved to just before the _New API_ section.

Thinking more... I wonder if this could be split into two blog posts? A first one that acts more as a 'quick start' introduction:

- What is this feature?
- Why should I be interested (what problems does it solve)?
- How do I get started?
- Quick next steps

And then a follow up that's a deeper dive into advanced examples and use-cases, how you built it, and what further work we want to do on it. What do you think?

Reply:
That said - I really like the post you've got here, maybe splitting it up is unnecessary extra work.

Reply:
Let's keep it like that, I think we'll need advanced cases blog post later anyway, but I don't have many real advanced cases yet.

Reply:
I've changed the ordering a bit though.